File: go/logkit/record.go
package logkit

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"
)

// SchemaVersion is the record contract's MAJOR, carried on every record so a
// line found on its own is self-describing.
const SchemaVersion = 1

// Fields is structured context for an event: the variable data that would
// otherwise be interpolated into Message. A key must not collide with a root
// field name; Record.Validate rejects the collision.
type Fields map[string]any

// Error is the log-side view of a failure. It is independent of Level and is
// not clikit's Error contract.
type Error struct {
	Message string   `json:"message"`
	Kind    string   `json:"kind,omitempty"`
	Stack   []string `json:"stack,omitempty"`
}

// Caller is the source location of a log call, module-relative.
type Caller struct {
	File     string `json:"file"`
	Line     int    `json:"line"`
	Function string `json:"function,omitempty"`
}

// Record is one log event in its normalized form: the exact field set of
// schemas/logkit/log-record.schema.json. Fields are declared in the schema's
// own alphabetical root order so a direct struct marshal already carries
// canonical key order at the root; Marshal still runs the result through an
// RFC 8785 canonicalizer for the nested values (fields, error, caller) and
// for numbers and escaping.
type Record struct {
	Caller         *Caller `json:"caller,omitempty"`
	Error          *Error  `json:"error,omitempty"`
	Fields         Fields  `json:"fields,omitempty"`
	Level          Level   `json:"level"`
	Message        string  `json:"message"`
	SchemaVersion  int     `json:"schema_version"`
	Service        string  `json:"service"`
	ServiceVersion string  `json:"service_version,omitempty"`
	Timestamp      string  `json:"timestamp"`
}

// FormatTimestamp renders t per the contract: UTC, truncated (never rounded)
// to the millisecond, exactly three fractional digits, RFC 3339 with a literal
// Z offset.
func FormatTimestamp(t time.Time) string {
	return t.UTC().Truncate(time.Millisecond).Format("2006-01-02T15:04:05.000Z")
}

// isLine reports whether s satisfies the schema's $defs/line: non-empty, at
// most 4096 characters, and free of control characters (so it survives both
// the one-line JSON rendering and the human rendering unchanged).
func isLine(s string) bool {
	if s == "" || len(s) > 4096 {
		return false
	}
	for _, r := range s {
		if r < 0x20 || r == 0x7f {
			return false
		}
	}
	return true
}

var rootFieldNames = map[string]bool{
	"caller": true, "error": true, "fields": true, "level": true,
	"message": true, "schema_version": true, "service": true,
	"service_version": true, "timestamp": true,
}

var servicePattern = func(s string) bool {
	if s == "" || len(s) > 64 {
		return false
	}
	first := s[0]
	if !(first >= 'a' && first <= 'z' || first >= '0' && first <= '9') {
		return false
	}
	for i := 0; i < len(s); i++ {
		c := s[i]
		ok := c >= 'a' && c <= 'z' || c >= '0' && c <= '9' || c == '.' || c == '_' || c == '-'
		if !ok {
			return false
		}
	}
	return true
}

// Validate checks r against the rules log-record.schema.json cannot express
// as types alone: the service pattern, line fields, the fields/root-name
// collision, and the presence of the required fields. It is the last gate
// before a record is canonicalized and emitted.
func (r *Record) Validate() error {
	if r.SchemaVersion != SchemaVersion {
		return fmt.Errorf("logkit: schema_version must be %d, got %d", SchemaVersion, r.SchemaVersion)
	}
	if !r.Level.Known() {
		return fmt.Errorf("logkit: unknown level %q", string(r.Level))
	}
	if !servicePattern(r.Service) {
		return fmt.Errorf("logkit: invalid service %q", r.Service)
	}
	if !isLine(r.Message) {
		return fmt.Errorf("logkit: invalid message %q", r.Message)
	}
	if r.Timestamp == "" {
		return fmt.Errorf("logkit: timestamp is required")
	}
	for key := range r.Fields {
		if key == "" || strings.ContainsAny(key, "\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x7f") {
			return fmt.Errorf("logkit: invalid fields key %q", key)
		}
		if rootFieldNames[key] {
			return fmt.Errorf("logkit: fields key %q collides with a root field name", key)
		}
	}
	// NaN and the infinities are not JSON and are rejected here, never coerced
	// (contract canonical_json.numbers). This also rejects any other value that
	// cannot serialize, which the byte writer would otherwise swallow into a
	// placeholder string.
	if len(r.Fields) > 0 {
		if _, err := json.Marshal(r.Fields); err != nil {
			return fmt.Errorf("logkit: invalid fields value: %w", err)
		}
	}
	if r.Error != nil && !isLine(r.Error.Message) {
		return fmt.Errorf("logkit: invalid error.message %q", r.Error.Message)
	}
	if r.Caller != nil {
		if !isLine(r.Caller.File) || strings.HasPrefix(r.Caller.File, "/") || strings.HasPrefix(r.Caller.File, `\`) {
			return fmt.Errorf("logkit: invalid caller.file %q", r.Caller.File)
		}
		if r.Caller.Line < 1 {
			return fmt.Errorf("logkit: invalid caller.line %d", r.Caller.Line)
		}
	}
	return nil
}

File: rust/logkit/src/record.rs
//! The normalized record: `log-record.schema.json` given a Rust shape.
//!
//! A [`Record`] is always schema-valid by construction — every path that
//! builds one goes through this module's validators — so nothing downstream
//! re-checks it against the schema at runtime. [`Record::canonical_json`]
//! and [`Record::to_value`] are the two ways a finished record leaves this
//! module: the wire form and the form [`crate::render`] renders for people.

use serde::Serialize;
use serde_json::Map;
use serde_json::Value;

use crate::error::LogError;
use crate::level::Level;

/// MAJOR of the record contract this crate emits. Per the schema: a
/// consumer built against this crate refuses a record declaring more.
pub const SCHEMA_VERSION: u32 = 1;

/// Every root field name, in the schema's own order. `fields` keys are
/// checked against this set so a context key can never shadow a record
/// field.
pub const ROOT_FIELD_NAMES: [&str; 9] = [
    "caller",
    "error",
    "fields",
    "level",
    "message",
    "schema_version",
    "service",
    "service_version",
    "timestamp",
];

/// The failure that caused the event, when one did. Independent of `level`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct ErrorInfo {
    /// The failure's own message, verbatim from the error value.
    pub message: String,
    /// The error's type as Rust names it. Absent when none is known.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub kind: Option<String>,
    /// Stack frames, outermost-call first. Never an empty array — omitted
    /// instead when there is nothing to capture.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stack: Option<Vec<String>>,
}

/// Source location of the log call.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct CallerInfo {
    /// Path from the module root — never absolute.
    pub file: String,
    /// 1-based line number of the log call.
    pub line: u32,
    /// Enclosing function or method, when the runtime can resolve one.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub function: Option<String>,
}

/// One log event in its normalized form.
///
/// Every field here mirrors `log-record.schema.json`'s `properties`
/// one-for-one, by name. Build one through [`crate::Logger`]'s event
/// builder, which runs the validators below before a `Record` ever exists;
/// there is no public constructor that skips them.
#[derive(Debug, Clone, Serialize)]
pub struct Record {
    /// MAJOR of the record contract; always [`SCHEMA_VERSION`].
    pub schema_version: u32,
    /// RFC 3339 UTC millisecond instant the event happened.
    pub timestamp: String,
    /// Severity, from the closed five-member set.
    pub level: Level,
    /// The logical emitter — the CLI, daemon or job name.
    pub service: String,
    /// What happened, as a short constant phrase.
    pub message: String,
    /// Version of the emitting build, when known.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub service_version: Option<String>,
    /// Structured context. Omitted entirely when empty, never `{}`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fields: Option<Map<String, Value>>,
    /// The failure that caused the event, when one did.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<ErrorInfo>,
    /// Source location of the log call, when captured.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub caller: Option<CallerInfo>,
}

impl Record {
    /// RFC 8785 (JCS) canonical serialization: sorted keys at every depth,
    /// no whitespace, ECMA-262 numbers. This is the wire form and the byte
    /// sequence three independent implementations agree on.
    ///
    /// # Errors
    /// Only if a `fields` value is structurally not JSON (this crate's own
    /// `serde_json::Value` values always are), so a caller sees this as
    /// effectively infallible.
    pub fn canonical_json(&self) -> Result<String, serde_json::Error> {
        serde_jcs::to_string(self)
    }

    /// The record as a `serde_json::Value`, keys sorted alphabetically
    /// (this crate never enables `serde_json`'s `preserve_order` feature, so
    /// every `Map` is a `BTreeMap` under the hood). [`crate::human_line`] walks
    /// this to derive the human line's attribute order — the same
    /// alphabetical root order the JCS canonicalizer produces for ASCII
    /// keys, so both renderings agree on "the record's own canonical key
    /// order" without deriving one from the other.
    ///
    /// # Panics
    /// Never in practice: every field on this type already serializes to
    /// JSON, so the underlying `serde_json` call cannot fail.
    #[must_use]
    pub fn to_value(&self) -> Value {
        serde_json::to_value(self).expect("Record serializes to a JSON object by construction")
    }
}

/// A non-empty, single-line, length-bounded string: no control character
/// (U+0000-U+001F or U+007F), so it survives both renderings unchanged.
/// Backs `message`, `error.message`, `error.kind`, each stack frame,
/// `caller.function` and `service_version`.
pub(crate) fn validate_line(
    field: &'static str,
    value: &str,
    max_len: usize,
) -> Result<(), LogError> {
    if value.is_empty() {
        return Err(invalid(field, value, "must not be empty"));
    }
    if value.chars().count() > max_len {
        return Err(invalid(field, value, "exceeds the maximum length"));
    }
    if value.chars().any(is_control) {
        return Err(invalid(
            field,
            value,
            "must not contain a control character",
        ));
    }
    Ok(())
}

/// `service`: `^[a-z0-9][a-z0-9._-]*$`, 1-64 characters.
pub(crate) fn validate_service(value: &str) -> Result<(), LogError> {
    let mut chars = value.chars();
    let reason = "must match ^[a-z0-9][a-z0-9._-]*$ and be 1-64 characters";
    match chars.next() {
        Some(c) if c.is_ascii_lowercase() || c.is_ascii_digit() => {}
        _ => return Err(invalid("service", value, reason)),
    }
    if value.chars().count() > 64
        || !chars
            .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || matches!(c, '.' | '_' | '-'))
    {
        return Err(invalid("service", value, reason));
    }
    Ok(())
}

/// `caller.file`: a control-character-free, module-relative path — no
/// leading `/` or `\`, since an absolute path bakes in the build machine's
/// directory layout and breaks byte-identity across hosts.
pub(crate) fn validate_caller_file(value: &str) -> Result<(), LogError> {
    validate_line("caller.file", value, 256)?;
    if value.starts_with('/') || value.starts_with('\\') {
        return Err(invalid(
            "caller.file",
            value,
            "must be relative, not absolute",
        ));
    }
    Ok(())
}

/// A `fields` key: non-empty, control-character-free, at most 128
/// characters, and not one of [`ROOT_FIELD_NAMES`] (the schema's
/// `not: root_field_name` rule — a context key can never shadow a root
/// field).
pub(crate) fn validate_field_key(key: &str) -> Result<(), LogError> {
    validate_line("fields key", key, 128)?;
    if ROOT_FIELD_NAMES.contains(&key) {
        return Err(LogError::ReservedFieldName {
            key: key.to_string(),
        });
    }
    Ok(())
}

fn is_control(c: char) -> bool {
    c.is_control()
}

fn invalid(field: &'static str, value: &str, reason: &'static str) -> LogError {
    LogError::InvalidValue {
        field,
        reason,
        value: value.to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn service_accepts_the_schema_pattern() {
        for ok in ["navigator", "git-tools", "anoikis-tools", "a", "a0._-z"] {
            validate_service(ok).unwrap();
        }
    }

    #[test]
    fn service_rejects_uppercase_leading_symbol_and_overlong() {
        for bad in ["Navigator", "-nav", "", &"a".repeat(65)] {
            assert!(
                validate_service(bad).is_err(),
                "expected {bad:?} to be rejected"
            );
        }
    }

    #[test]
    fn line_rejects_empty_and_control_characters() {
        assert!(validate_line("message", "", 4096).is_err());
        assert!(validate_line("message", "line one\nline two", 4096).is_err());
        validate_line("message", "index rebuilt", 4096).unwrap();
    }

    #[test]
    fn field_key_rejects_root_field_collision() {
        assert!(validate_field_key("service").is_err());
        assert!(validate_field_key("documents").is_ok());
    }

    #[test]
    fn caller_file_rejects_absolute_paths() {
        assert!(validate_caller_file("/abs/path.rs").is_err());
        validate_caller_file("rust/bm25/src/index.rs").unwrap();
    }

    #[test]
    fn empty_optional_fields_are_omitted_not_null_or_empty() {
        let record = Record {
            schema_version: SCHEMA_VERSION,
            timestamp: "2026-07-26T09:41:07.480Z".to_string(),
            level: Level::Info,
            service: "navigator".to_string(),
            message: "index rebuilt".to_string(),
            service_version: None,
            fields: None,
            error: None,
            caller: None,
        };
        let json = record.canonical_json().unwrap();
        assert!(!json.contains("fields"));
        assert!(!json.contains("null"));
        assert_eq!(
            json,
            r#"{"level":"info","message":"index rebuilt","schema_version":1,"service":"navigator","timestamp":"2026-07-26T09:41:07.480Z"}"#
        );
    }
}
